packages/dagu_car_dbv2/src/kinematics_node.py

```diff
@@ -0,0 +1,327 @@
+#!/usr/bin/env python
+
+import rospy
+import yaml
+import time
+import os.path
+
+from duckietown import DTROS
+from duckietown_msgs.msg import WheelsCmdDBV2Stamped, Twist2DStamped
+from std_srvs.srv import EmptyResponse, Empty
+import math
+
+class KinematicsNode(DTROS):
+    """
+    The `KinematicsNode` maps car commands send from various nodes to wheel commands
+    that the robot can execute.
+
+    The `KinematicsNode` performs both the inverse and forward kinematics calculations. Before
+    these were implemented in separate nodes, but due to their similarity and parameter sharing,
+    they are now combined.
+
+    `KinematicsNode` utilises the car geometry as well as a number of tunining and limiting
+    parameters to calculate the wheel commands that the wheels should execute in order for the
+    robot to perform the desired car commands (inverse kinematics). Then it uses these wheel
+    commands in order to do an open-loop velocity estimation (the forward kinematics part).
+
+    TODO: Add link/explanation/illustration of the car geometry and the math used
+
+    All the configuration parameters of this node can be changed dynamically while the node
+    is running via `rosparam set` commands.
+
+    Args:
+        node_name (:obj:`str`): a unique, descriptive name for the node that ROS will use
+
+    Configuration:
+        ~gain (:obj:`float`): scaling factor applied to the desired velocity, default is 1.0
+        ~trim (:obj:`float`): trimming factor that is typically used to offset differences in the
+            behaviour of the left and right motors, it is recommended to use a value that results in
+            the robot moving in a straight line when forward command is given, default is 0.0
+        ~baseline (:obj:`float`): the distance between the two wheels of the robot, default is 0.1
+        ~radius (:obj:`float`): radius of the wheel, default is 0.0318
+        ~k_wheel (:obj:`float`): motor constant, assumed equal for both motors, default is 27.0
+        ~limit (:obj:`float`): limits the final commands sent to the motors, default is 1.0
+        ~v_max (:obj:`float`): limits the input velocity, default is 1.0
+        ~omega_max (:obj:`float`): limits the input steering angle, default is 8.0
+
+    Subscriber:
+        ~car_cmd (:obj:`Twist2DStamped`): The requested car command
+
+    Publisher:
+        ~wheels_cmd_dbv2 (:obj:`WheelsCmdDBV2Stamped`): The corresponding resulting wheel commands
+        ~velocity (:obj:`Twist2DStamped`): The open-loop estimation of the robot velocity
+
+    Service:
+        ~save_calibration:
+            Saves the current set of kinematics parameters (the ones in the Configuration section)
+                to `/data/config/calibrations/kinematics/HOSTNAME.yaml`.
+
+    """
```

Comment:
Is this whole description up to date with the dbv2 kinematics ?

Reply:
I don't know. @hosnerm ?